src/Servers/HttpSys/src/HttpSysListener.cs

```diff
@@ -310,103 +309,78 @@ internal unsafe void SendError(ulong requestId, int httpStatusCode, IList<string
+            {
+                // Add Value
+                string headerValue = authChallenges[headerValueIndex];
+                bytes = allocator.GetHeaderEncodedBytes(headerValue, out bytesLength);
+                nativeHeaderValues[header->KnownHeaderCount].RawValueLength = (ushort)bytesLength;
+                nativeHeaderValues[header->KnownHeaderCount].pRawValue = bytes;
+                header->KnownHeaderCount++;
             }
 
-            httpResponse.Response_V1.StatusCode = (ushort)httpStatusCode;
-            string? statusDescription = HttpReasonPhrase.Get(httpStatusCode);
-            uint dataWritten = 0;
-            uint statusCode;
-            byte[] byteReason = statusDescription != null ? HeaderEncoding.GetBytes(statusDescription) : Array.Empty<byte>();
-            fixed (byte* pReason = byteReason)
-            {
-                httpResponse.Response_V1.pReason = (byte*)pReason;
-                httpResponse.Response_V1.ReasonLength = (ushort)byteReason.Length;
+            knownHeaderInfo[0].pInfo = header;
 
-                byte[] byteContentLength = new byte[] { (byte)'0' };
-                fixed (byte* pContentLength = byteContentLength)
-                {
-                    (&httpResponse.Response_V1.Headers.KnownHeaders)[(int)HttpSysResponseHeader.ContentLength].pRawValue = (byte*)pContentLength;
-                    (&httpResponse.Response_V1.Headers.KnownHeaders)[(int)HttpSysResponseHeader.ContentLength].RawValueLength = (ushort)byteContentLength.Length;
-                    httpResponse.Response_V1.Headers.UnknownHeaderCount = 0;
-
-                    statusCode =
-                        HttpApi.HttpSendHttpResponse(
-                            _requestQueue.Handle,
-                            requestId,
-                            0,
-                            &httpResponse,
-                            null,
-                            &dataWritten,
-                            IntPtr.Zero,
-                            0,
-                            SafeNativeOverlapped.Zero,
-                            IntPtr.Zero);
-                }
-            }
-            if (statusCode != UnsafeNclNativeMethods.ErrorCodes.ERROR_SUCCESS)
-            {
-                // if we fail to send a 401 something's seriously wrong, abort the request
-                HttpApi.HttpCancelHttpRequest(_requestQueue.Handle, requestId, IntPtr.Zero);
-            }
+            httpResponse.ResponseInfoCount = 1;
         }
-        finally
+
+        httpResponse.Response_V1.StatusCode = (ushort)httpStatusCode;
+        string statusDescription = HttpReasonPhrase.Get(httpStatusCode) ?? string.Empty;
+        uint dataWritten = 0;
+        uint statusCode;
+
+        bytes = allocator.GetHeaderEncodedBytes(statusDescription, out bytesLength);
+        httpResponse.Response_V1.pReason = bytes;
+        httpResponse.Response_V1.ReasonLength = (ushort)bytesLength;
```

Comment:
Given that the length is passed around everywhere, does httpsys require the header-encoded bytes need to be zero-terminated? We can save a bit by skipping the zero-termination.

Reply:
They do not. I am simply terminating because that helps with viewing the buffer. I prefer terminating as a matter of consistency in the buffer logic.